### ISSUE TITLE
Netboot does not work due to UDP-mapping

### ISSUE THREAD

**Issue by neerdoc** [linuxserver/docker-netbootxyz#11]
[linuxserverurl]: https://linuxserver.io
[![linuxserver.io](https://raw.githubusercontent.com/linuxserver/docker-templates/master/linuxserver.io/img/linuxserver_medium.png)][linuxserverurl]

If you are new to Docker or this application our issue tracker is **ONLY** used for reporting bugs or requesting features. Please use [our discord server](https://discord.gg/YWrKVTn) for general support.

------------------------------

## Expected Behavior
After starting the docker containers it should be possible to do a netboot.

## Current Behavior
Netboot fails due to NFTP connection over UDP is only initialized but fails during handshake since it tries to open new random UDP-ports which are not mapped through to the docker containers.
### Working example
Add `network_mode: host` to the docker-compose.yaml file. This let's the container utilize a bridged network interface which enable the TFTP handshake to succeed.

## Steps to Reproduce

1. Start setup by running `docker-compose up`
2. Start a virtualbox (or similar) and try netbooting.
3. Booting tries to download the `netboot.xyz.kpxe` through TFTP but fails with a timeout.

## Environment
**OS:**     
**CPU architecture:** x86_64/arm32/arm64      
**How docker service was installed:**

**Comment by thelamer** [comment]
I think you are jumping to conclusions as to why you are having a problem. 
This container has 3 million pulls and outside of my local deployment I know of hundreds of other functional deployments. 
Your troubleshooting and conclusion would indicate that this container is non functional which is not true.

In any case I do not believe this is a bug report or feature request, please hop on netboot.xyz or linuxsever discord/discourse for help setting up this container in your specific network.

**Comment by ProfessorManhattan** [comment]
@neerdoc Thank you so much for providing this solution! I get to go to sleep now.

It seems like this issue might be dependent on the router/router settings. When I was using UniFi UDM Pro, I just followed the instructions. However, with my slightly less forgiving pfSense set up, the only thing that got it working was adding this setting (and updating the docker-compose version number to 3.7).

**Comment by ProfessorManhattan** [comment]
@thelamer I think this might have something to do with PnP settings on the router... not sure though. If there really are millions of users, I'm sure plenty of them (especially ones with strict firewall settings) are going to have this issue... I think you should add it to the documentation

**Comment by vangorra** [comment]
After looking around a bit, I learned that TFTP establishes connections over port 69 UDP but transfers occur over an ephemeral port. I tried setting the UDP port range map in the docker config (which is used by in.tftpd) but transfers would still timeout. I'm not sure why this happens. At any rate, I worked around the issue by using the following docker compose file. This works around the issue by allowing tftp to run in host network mode (so all ports are exposed) and netboot is still using bridged.
```
version: "3"

# More info at https://blog.linuxserver.io/2019/12/16/netboot-xyz-docker-network-boot-server-pxe/ and https://docs.linuxserver.io/images/docker-netbootxyz
services:
  netbootxyz:
    image: linuxserver/netbootxyz:latest
    ports:
      - "3000:3000" # HTTP admin interface
      - "8080:80" # HTTP serve
    environment:
      PUID: 1026
      PGID: 100
      PORT_RANGE: 30000:30010
    volumes:
      - './container_data/netbootxyz/config:/config'
      - './container_data/netbootxyz/assets:/assets'
    restart: unless-stopped
  tftpd:
    image: stoxygen/tftpd-hpa:v1.0.0
    network_mode: 'host' # TFTP file transfers don't work without host network mode.
    volumes:
      - './container_data/netbootxyz/config/menus:/data'
    restart: unless-stopped
```

**Comment by ProfessorManhattan** [comment]
@lpyparmentier are you sure those are the only required ports? I thought it used any ephemeral port

**Comment by lpyparmentier** [comment]
Not at all :)

**Comment by lpyparmentier** [comment]
In fact, I think i've talk a little bit too fast... sorry

I've installed `tftpd-hpa` to replace the tftpd service provided by the container. And it works locally, but not on other machines. So I came back with the docker-compose provided by the documentation and I'm stuck with the same issue that having having the service without docker.

For now all i'm seeing is that i've bad udp chksum on the ephemeral port (data exchange):
```sh
root@rockpi4-services:/mnt/nvme/netboot# tcpdump udp port 32769 -vv -X
tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
19:38:56.812520 IP (tos 0x0, ttl 64, id 32299, offset 0, flags [none], proto UDP (17), length 544)
    192.168.4.146.39184 > 192.168.4.126.32769: [bad udp cksum 0x8c7e -> 0x084f!] UDP, length 516
        0x0000:  4500 0220 7e2b 0000 4011 7041 c0a8 0492  E...~+..@.pA....
        0x0010:  c0a8 047e 9910 8001 020c 8c7e 0003 0001  ...~.......~....
        0x0020:  ea05 00c0 0766 9c66 600f a80f a006 1e66  .....f.f`......f
        0x0030:  684c 5245 542e 8c16 2105 2e66 8926 1d05  hLRET...!..f.&..
```

And no data on the router, so everything happen on the switch (layer 2). No idea how to deal with that.

**Comment by phockett** [comment]
Thanks @vangorra , this fix worked for me too. 

@lpyparmentier, @ProfessorManhattan  I'm still quite new to Docker, but agreed that it seems that the original docker-compose configuration would never work otherwise, unless Docker defaults to `host` networking in some/many/most cases(?). For reference, I'm running the container on a Synology NAS (DSM 6.2.4-25556, with Docker 20.10.3-0554) and had exactly the same tftp timeout at `netboot.xyz.kpxe` loading as the OP. My router is an Asus RT-AC68U with Merlin firmware, and configured for TFTP/PXE as per the [docker-netbootxyz DD-WRT guide](https://github.com/linuxserver/docker-netbootxyz#dd-wrt).

> I tried setting the UDP port range map in the docker config (which is used by in.tftpd) but transfers would still timeout.

I didn't dig into the port mappings in detail to see which ephemeral ports were used in the end in my case, but did quickly test just mapping the additional ports specified by `PORT_RANGE` in the compose file - this *didn't* fix the problem, although presumably (?) there should be a way to enforce this within the container tftp instance.

@thelamer Kudos for maintaining all this... is there something else I could provide to help debug this since, as you point out, it seems to work without issue for many other people...?

**Comment by crahan** [comment]
Here's how I worked around this issue on my Synology. The easiest fix is to simply configure the builtin Synology TFTP server to serve its files from the shared folder where the `netboot.xyz` docker menu files are located.

In my case I configured the docker image to store `config` and `assets` in `/docker/netbootxyz/config` and `/docker/netbootxyz/assets`, respectively. I didn't expose UDP port 69 on the docker container (since I want the Synology to handle that), and finally pointed the builtin Synology TFTP server to serve its content from `/docker/netbootxyz/config/menus`. That's all that was needed to get everything working.

Now the Synology takes care of all the TFTP stuff and the netboot.xyz webapp is used to manage the app (via exposed port 3000) and serve local assets (via exposed port 8080). No need to use host networking. Bridge networking works just fine with this setup.

**Comment by crahan** [comment]
@phockett here's how I worked around this issue on my Synology. The easiest fix is to simply configure the builtin Synology TFTP server to serve its files from the shared folder where the netboot.xyz docker menu files are located.

In my case I configured the docker image to store config and assets in /docker/netbootxyz/config and /docker/netbootxyz/assets, respectively. I didn't expose UDP port 69 on the docker container (since I want the Synology to handle that), and finally pointed the builtin Synology TFTP server to serve its content from /docker/netbootxyz/config/menus. That's all that was needed to get everything working.

Now the Synology takes care of all the TFTP stuff and the netboot.xyz webapp is used to manage the app (via exposed port 3000) and serve local assets (via exposed port 8080). No need to use host networking. Bridge networking works just fine with this setup.

**Comment by phockett** [comment]
Thanks @crahan - definitely also a viable solution/work-around/configuration, although doesn't address the underlying issue with the docker-compose configuration of course... and the fact/hope that Docker should make this super-simple, and platform independent, on the user end. (Probably some wishful thinking on my part there ;) )

**Comment by SDShooter** [comment]
TL;DR; - Network mode: host solves the ephemeral port problem if you're having one, and don't trust windows TFTP client.

I was pulling my hair out over this issue.  Running docker in an Ubuntu VM on a Windows 11 host through either Hyper-V or VMWare player using external networking (VM is able to share host network adapter and get it's own IP address).

I was using the built in TFTP client in Windows (11), and it would time out every time.  After using WireShark to do a packet capture, it appeared that the initial file request succeeded and the TFTPD server kept sending back Block 1 of the file over and over.  The reply from the client was missing, which would have signaled the server to send Block 2, Block 3, etc.

However, it looks like this might be some sort of incompatibility between tftpd-ha and the windows built in TFTP client as I tried to accomplish a PXE boot at this point at it worked as expected with network_mode: host defined in the docker compose file.

Anyways, thanks for this tip neerdoc. Without it I would continue to be Pixieless.

**Comment by robsel-0** [comment]
My solution to get netboot.xyz to work with rootless podman in bridged network is this: `TFTPD_OPTS=--tftp-single-port`
This option forces TFTP to use only port 69 within the container.

I use iptables to forward host port 69 to the container.


compose.yml:
```
services:
  netbootxyz:
    image: ghcr.io/netbootxyz/netbootxyz
    container_name: netbootxyz
    environment:
      - MENU_VERSION=2.0.84 # optional
      - NGINX_PORT=80 # optional
      - WEB_APP_PORT=3000 # optional
      - TFTPD_OPTS=--tftp-single-port
    volumes:
      - ../data/netboot.xyz/config:/config # optional
      - ../data/netboot.xyz/assets:/assets # optional
    ports:
      - 3100:3000  # optional, destination should match ${WEB_APP_PORT} variable above.
      - 6969:69/udp
      - 8083:80  # optional, destination should match ${NGINX_PORT} variable above.
    restart: unless-stopped

```
